src/lib/bootstrap.ts

```diff
@@ -104,24 +105,74 @@ function logHTTPSWarning(storageLocation) {
   process.exit(2);
 }
 
-function handleHTTPS(app, configPath, config) {
-  try {
-    let httpsOptions = {
-      secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3, // disable insecure SSLv2 and SSLv3
-    };
+let secureContext: tls.SecureContext | undefined;
+function handleSNICallback(domain, callback): tls.SecureContext | undefined {
+  if (secureContext) {
+    if (callback) {
+      callback(null, secureContext);
+    } else {
+      return secureContext
+    }
+  } else {
+    logger.logger.warn('No secureContext found');
+  }
+}
 
+function refreshSecureContext(config, nothrow?): void {
+  try {
+    let newContext : tls.SecureContext | undefined;
     if (config.https.pfx) {
-      httpsOptions = assign(httpsOptions, {
+      newContext = tls.createSecureContext({
         pfx: fs.readFileSync(config.https.pfx),
-        passphrase: config.https.passphrase || '',
+        passphrase: config.https.passphrase
       });
     } else {
-      httpsOptions = assign(httpsOptions, {
+      newContext = tls.createSecureContext({
         key: fs.readFileSync(config.https.key),
         cert: fs.readFileSync(config.https.cert),
         ca: fs.readFileSync(config.https.ca),
       });
     }
+    secureContext = newContext;
+    logger.logger.info('Secure context renewed');
+  } catch (err) {
+    logger.logger.warn({ err: err }, 'Cannot create/renew secure context: @{err.message}');
+    if (!nothrow) {
+      throw err;
+    }
+  }
+}
+
+function handleHTTPS(app, configPath, config) {
+  try {
+    refreshSecureContext(config);
+    const fileToWatch = config.https.pfx
+      ? config.https.pfx
+      : config.https.cert
+      ;
+
+
+    // add a filewatcher on the certificate files, so the verdaccio
+    // doesn't need to be restarted when the certificate changes 
+    fs.watch(
+      fileToWatch, 
+      (eventtype, filename): void => {
+        // add a bit of a timeout, to handle the (unlikely) case, that a cert-file is already changed, 
+        // but the correponding key-file is not yet updated
+        if (eventtype === 'change') {
+          setTimeout(() => {
+              refreshSecureContext(config, true);
+            }, 
+            5000); 
+          }
+      }
+    );
+
+    const httpsOptions = {
+      secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3, // disable insecure SSLv2 and SSLv3
+      SNICallback: handleSNICallback,
```

Comment:
When would `SNICallback` to be invoked? I don't know if it's for updating `secureContext` though, it would be great if you could share more detail!

By the way, after node.js v11, we can [`setSecureContext`](https://nodejs.org/api/tls.html#tls_server_setsecurecontext_options)

Reply:
`SNIcallback` will be executed every time a request via a a secure connection is made. `SNICallback` is not for updating the secure context, but for returning the correct secure context in an SNI situation (ie where multiple domains are pointing to the same server and you don't have a single certificate containing all of these domains)

The secure context is refreshed only on the event of the filewatcher, when the certificate file changed.

Yes, for the use case of verdaccio `setSecureContext` may be applicable, because you only have a single certificate. I took this code from another project of mine, where I had multiple certificates and which originated from nodejs v8, where this was the only way to replace the certificate without interruption of the connection

Reply:
> SNIcallback will be executed every time a request via a a secure connection is made. SNICallback is not for updating the secure context, but for returning the correct secure context in an SNI situation (ie where multiple domains are pointing to the same server and you don't have a single certificate containing all of these domains)

Thanks for your reply @wiggisser! You explained SNI situation very well!

After some research on this topic, I figured probably restarting the server after certificate renewal isn't in the scope of verdaccio itself. 

It's more related to how you renew your certificate. Take certbot for example, it has a `--nginx` flag which helps restarting nginx server after certificate renewal. In case a node.js server is running instead, it should restart within the command to renew the certificate in cron job or use [deploy-hook](https://certbot.eff.org/docs/using.html)

Reply:
@ddhp 
I would say, that depends. For verdaccio it may be enough to just restart the service, because it's startup time is generally quite short, and it may not be a big problem, if a few calls fail during the restart. For other services which may have quite a startuptime, it may be not acceptable to have a lengthy downtime just for certificate renewal. Or there may be some longrunning processes in those services, which should not be interrupted just for a certificate renewal. Just killing and restaring a service could result in some serious problems here. 

Again, for verdaccio it might be ok to just restart. Nonetheless I liked the idea of the service just to use the new certificate when it's available, without needing any intervention from the outside. But to me, it doesn't seem this is gonna be merged anytime soon. So restarting verdaccio or setting up a reverse-proxy might be inevitable anyway.